### PR TITLE
Make name optional

### PR DESCRIPTION
**Pull request by samtstern**
This `fixes` #467

Also I uncovered a **TERRIBLE** error from my previous change.  In `SaveSmartLock` I was switching the `name` and the `password`.  Gulp.  Should probably add a test for that.

### CODE REVIEW

File: auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SaveSmartLock.java [hunk 1]
```diff
@@ -142,7 +142,7 @@ public void saveCredentialsOrFinish(FirebaseUser firebaseUser,
         // Build credentials client and kick off the save
         mCredentialsClient = Credentials.getClient(getActivity());
         Credential credential = CredentialsUtil.buildCredential(
-                mEmail, mName, mPassword, mProfilePictureUri, mResponse);
+                mEmail, mPassword, mName, mProfilePictureUri, mResponse);
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SaveSmartLock.java]: This is the scariest thing.  Need to figure out what test should have failed and make it MUCH LOUDER.

**Comment by samtstern** [comment]
Ok I totally overhauled the `verifySmartLockSave` test helper so that it will detect future failures like the one I found manually.

File: auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java [hunk 2]
```diff
@@ -96,18 +101,25 @@ public View onCreateView(LayoutInflater inflater,
 
         View v = inflater.inflate(R.layout.fui_register_email_layout, container, false);
 
+        // Get configuration
+        AuthUI.IdpConfig emailConfig = ProviderUtils.getConfigFromIdps(
+                getFlowParams().providerInfo, EmailAuthProvider.PROVIDER_ID);
+        mRequireName = emailConfig.getParams().getBoolean(ExtraConstants.EXTRA_REQUIRE_NAME, true);
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailFragment.java]: Can we make this a local variable? Adding a field with an extra default value scares me. 😄

File: auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SaveSmartLock.java [hunk 3]
```diff
@@ -107,6 +108,20 @@ private void finish() {
         finish(Activity.RESULT_OK, mResponse.toIntent());
     }
 
+    @VisibleForTesting
```
Comment (SUPERCILEX) [inline auth/src/main/java/com/firebase/ui/auth/util/signincontainer/SaveSmartLock.java]: `    @RestrictTo(RestrictTo.Scope.TESTS)`?

File: auth/src/test/java/com/firebase/ui/auth/ui/email/EmailActivityTest.java [hunk 4]
```diff
@@ -91,13 +90,16 @@ public void testSignUpButton_validatesFields() {
                                 .getInteger(R.integer.fui_min_password_length)
                 ),
                 passwordLayout.getError().toString());
+
+
```
Comment (SUPERCILEX) [inline auth/src/test/java/com/firebase/ui/auth/ui/email/EmailActivityTest.java]: Oops?

**Pull request merged**